kernel/src/common/buffer.rs

```diff
@@ -0,0 +1,129 @@
+//! Defines a LeasableBuffer type which can be used to pass a section of a larger
+//! buffer but still get the entire buffer back in a callback
+//!
+//! Author: Amit Levy
+//!
+//! Usage
+//! -----
+//!
+//! `slice()` is used to set the portion of the `LeasableBuffer` that is accessbile.
+//! `reset()` makes the entire `LeasableBuffer` accessible again.
```

Comment:
Does anything prevent the lower layer from calling `reset()`?

Reply:
Currently, no. This design is more to protect against accidental misuse of the whole buffer. Prior to using this construct we would always pass the entire buffer along with the length that the lower layer was actually supposed to access, but this was more mistake prone

Reply:
Ok gotcha. I'm wondering if `slice()` could return something that must be passed to `reset()` and that can't be created by the thing that got the sliced buffer.

Reply:
But that wouldn't be in this PR.